### PR TITLE
[flutter_local_notifications]: allow to set largeIcon from base64 encoded string and byte array

### PR DESCRIPTION
**Pull request by lexxxel**
Basically, this allows to set the large icon from flutter directly, e.g. when the image is received from a firebase cloud message.

### CODE REVIEW

**Comment by lexxxel** [comment]
This PR goes into the direction of #479 and #1122, while changing the API as little as possible.

File: flutter_local_notifications/lib/src/platform_specifics/android/enums.dart [hunk 1]
```diff
@@ -91,8 +94,7 @@ class Importance {
   static const Importance max = Importance(5);
 
   /// All the possible values for the [Importance] enumeration.
-  static List<Importance> get values =>
-      <Importance>[unspecified, none, min, low, defaultImportance, high, max];
+  static List<Importance> get values => <Importance>[unspecified, none, min, low, defaultImportance, high, max];
```
Comment (lexxxel) [inline flutter_local_notifications/lib/src/platform_specifics/android/enums.dart:97]: It seems that my android studio did not pickup your formatting style. Where do I find the config for the line with?

**Comment by lexxxel** [comment]
It seems that I missed something, BigPicture notifications do not work, yet

**Comment by lexxxel** [comment]
Fixed Big Picture Notifications and added this feature to the example app

**Comment by MaikuB** [comment]
Thanks for the PR. According to the official docs at https://flutter.dev/docs/development/platform-integration/platform-channels#codec, you should be able to pass Uint8List from the Dart side and that would be a byte array on the Java end. Would you able to take a look at that?

**Comment by lexxxel** [comment]
Yes, I saw the byte list conversion, too. I think base64 is a valid use case (e.g. I send the image inside the FCM data map<string,string>, so my image is already base64) and it was a simple extension of your code. I think Uint8List would require a refactoring. I might do that, still, but as a future PR.

**Comment by MaikuB** [comment]
I agree base64 is valid though in my experience, it tends to be used in cases like this because the transport mechanism won't allow for a byte array. Which leads to me the point I was trying to communicate where if using a byte array is possible, then I would say that is the way to go. Its more generic and and there a more use cases that can make use of it directly. This is in also demonstrated in the changes done in the example app where the response from downloading a file over the web leads to a byte array response. It then ends up converting it to a byte array on the Android side as well. Same applies for opening a file using Dart. With support for byte arrays, it would also make exposing an API of sorts for base64 redundant.

Is there something that prevents making the change in this PR? I know there would be some changes needed on the Android side, which I'm happy to pick up and assist with by carrying on from where this PR is at when time permits

**Comment by lexxxel** [comment]
Sounds good. I think this PR is ready to be reviewed/merged. Let's work on the byte array afterwards.

**Review by vytautas-pranskunas- — APPROVED** [review]

**Comment by MaikuB** [comment]
@lexxxel sorry perhaps I wasn't clear enough with my previous response. What I was referring to was to either change this PR to only support byte array or this can be closed and a new one be created to do so.

**Comment by lexxxel** [comment]
Sorry that I misunderstood you. Would you agree to support both, Base64 strings as well as byte arrays, but with the base64 string beeing converted to byte arrays before being passed to Android? 
Base64 strings are still useful when using sqlite.

**Comment by vytautas-pranskunas-** [comment]
HI @lexxxel, I was also thinking more about thi tonight. So even if you plan to introduce Url support with byte array in the future, today it would be usefule to introduce Url support wich under the hood will be converted to base64 and passed to notification. Why? because for user like me, who will be using Url, in upcomming versions library API will not change after you will swap base64 usage to bytes array.

**Comment by vytautas-pranskunas-** [comment]
@lexxxel If you an improve a bit PR ASAP it would be great
@MaikuB I would really appreciate if you can merge this PR ASAP because we have release in few upcoming days and this feature would be so benefitial for us :)

**Comment by lexxxel** [comment]
@MaikuB the statement from @vytautas-pranskunas- is true. I would also like to go this way. Adding URL support now and switch to byte arrays later

**Comment by lexxxel** [comment]
@MaikuB the current version does not use byte arrays to pass the data to java, but the 'interface' is ready to do that in a next step, do you agree?

**Comment by MaikuB** [comment]
> Would you agree to support both, Base64 strings as well as byte arrays, but with the base64 string being converted to byte arrays before being passed to Android?

Adding it via the use of a named constructor like you've done in the PR is a good idea. However, the way URL is support has a couple of issues IMO so I'd like to see it removed

- the API looks follows the naming convention of a named constructor. However, constructors are typically synchronous but this method is asynchronous
- a dependency on `http` has been added. This is the main reason to do so. This helps minimise dependencies and me other developers will may also use different libraries for network requests like Dio. With byte array support added, developers can use whichever library they want and would have the byte array they need to pass to this plugin. Using `http` is also problematic as it has not reached version 1.0 or above. Specifying `^0.13.3` in pubspec means if 0.14.0 is released, this plugin *wouldn't* grab it as a dependency. This ends up causing dependency conflicts and more issues being opened. This is something I've experienced with another library using rxdart, which is less commonly used even.

>  the current version does not use byte arrays to pass the data to java, but the 'interface' is ready to do that in a next step, do you agree?

I don't fully understand what you mean but I can see that the Dart class still holds the data using a string as opposed to a Uint8List. The `ByteArrayAndroidBitmap.fromBase64String` is missing some logic to [decode](https://api.dart.dev/stable/2.12.4/dart-convert/base64Decode.html). Not sure if this was forgotten, if you were looking for feedback before proceeding or was trying to see if I could carry with the next step (which I'm happy to assist with btw)

**Comment by lexxxel** [comment]
> Not sure if this was forgotten, if you were looking for feedback before proceeding or was trying to see if I could carry with the next step (which I'm happy to assist with btw)

Correct, I was looking for feedback.
I will remove the URL part, but show in the example how one could achieve the same thing. 

I will now continue with the refactoring to get all Uint8List based. `ByteArrayAndroidBitmap.fromBase64String` will then get some logic 😉

**Comment by vytautas-pranskunas-** [comment]
@lexxxel hi, how much more time do you think you need to complete this task? :)

**Comment by lexxxel** [comment]
> @lexxxel hi, how much more time do you think you need to complete this task? :)

I will be able to work on this by the end of this week or early next week. But this is not a promise.

**Comment by vytautas-pranskunas-** [comment]
I have tried to refactor a bit but have stucked with Java part...
However I have noticed one improvment that should be made also.
```largeIcon``` now is part of ```BigPictureStyleInformation``` where big picture is mandatory and largeIcon is optional. We need to make either new type of information eithe make big picture optional too because now user will not be able to show large avatar without large picture under :(

**Comment by lexxxel** [comment]
You can set `largeIcon` on 'normal' notifications. Why would you try to use big picture of you do not want a big picture?

**Comment by vytautas-pranskunas-** [comment]
@lexxxel ok, i saw it only on big picture in your examples :) if it is available there then np. Waiting for your code refactorings :)

**Comment by vytautas-pranskunas-** [comment]
@lexxxel @MaikuB Is it ready to be used?

**Comment by lexxxel** [comment]
@MaikuB I think its ready to be reviewed, especially `hasInvalidLargeIcon` and `hasInvalidBigPictureResources`. I honestly do not like what happens there, even before my refactoring. Maybe @MaikuB you have a good Idea how to make that part simpler.

**Comment by vytautas-pranskunas-** [comment]
@lexxxel @MaikuB I would love to see it merged and create future task for simplifying andrefactoring some parts because it i snot last release. Incremental improvements are very welcome hovever too long polishing leads to long release cycles. What do you think?

**Comment by lexxxel** [comment]
@vytautas-pranskunas- a bugged release with an unstable API does not help anyone, even you. 

If you are in a hurry, use my repo and branch. However,  if you are in testing right now, this should mean you are already behind your feature freeze, right? So maybe you should add this feature to your project in a future release 🙃🙂, 'because it is not last release' ☝️

**Comment by MaikuB** [comment]
@vytautas-pranskunas- I second what @lexxxel said. If you're in urgent need of using it then you can point your own app to the branch @lexxxel has or fork it and then point your app to it

@lexxxel thanks, will take a look when I get a chance to

**Comment by vytautas-pranskunas-** [comment]
I completely agree about buggy software, however improving code quality not fixing bugs 😉
But I hear what your are saying.
Anyway - how do you think how long it will take to approve this PR?

**Comment by vytautas-pranskunas-** [comment]
BTW question - is this avatar supported by iOS? I see only android in example?

**Comment by vytautas-pranskunas-** [comment]
@lexxxel would it be also possible to have some kind of builder to make largeIcon rounded (like avatar)

**Comment by lexxxel** [comment]
> @lexxxel would it be also possible to have some kind of builder to make largeIcon rounded (like avatar)?
> Or maybe to add predefined shapes...
> 
> example here: https://stackoverflow.com/questions/28280615/make-notification-large-icon-round

Of course, you can do that in flutter, or on your server that delivers the images. If you want that feature, simply create a PR 👍

**Comment by vytautas-pranskunas-** [comment]
@lexxxel in flutter? You mean inside of this plugin using JAVA? Because from flutter side you are passing byte array.
From server - my server is amazon S3 :)))
I know this is ot of scope but sadly I am not JAVA guy not even know how to run it thats why i stucked when trying to improve this plugin :(

**Comment by vytautas-pranskunas-** [comment]
@lexxxel - I know this out of scope but you can include this to your code as it anyway ment for largeIcon :)

```
static toRoundedCorners(Uint8List src,
      {int size = 100, bool addBorder = false, Color borderColor = Colors.white, double borderSize = 10}) async {
    final pictureRecorder = PictureRecorder();
    final Canvas canvas = Canvas(pictureRecorder);
    final Paint paint = Paint()..color;
    final double radius = size / 2;

    //make canvas clip path to prevent image drawing over the circle
    final Path clipPath = Path();
    clipPath
        .addRRect(RRect.fromRectAndRadius(Rect.fromLTWH(0, 0, size.toDouble(), size.toDouble()), Radius.circular(200)));
    canvas.clipPath(clipPath);

    //paintImage
    final Codec codec = await instantiateImageCodec(src);
    final FrameInfo imageFI = await codec.getNextFrame();
    paintImage(canvas: canvas, rect: Rect.fromLTWH(0, 0, size.toDouble(), size.toDouble()), image: imageFI.image);

    if (addBorder) {
      //draw Border
      paint..color = borderColor;
      paint..style = PaintingStyle.stroke;
      paint..strokeWidth = borderSize;
      canvas.drawCircle(Offset(radius, radius), radius, paint);
    }

    //convert canvas as PNG bytes
    final _image = await pictureRecorder.endRecording().toImage(size, size);
    final data = await _image.toByteData(format: ImageByteFormat.png);

    //convert PNG bytes as BitmapDescriptor
    return data!.buffer.asUint8List();
  }
```

**Comment by lexxxel** [comment]
@vytautas-pranskunas- out of scope! You answered many questions yourself or are able to figure out how to push circular images to your s3. For clarification, I meant flutter and dart, not Java.

This PR targets Android only. The iOS API of this lib does not offer a largeIcon as far as I know.

**Comment by vytautas-pranskunas-** [comment]
@lexxxel while testing it on my project I have found that when I am trying to use largIcon on zonedSchedule app crashes with error: 

```
/AndroidRuntime(13396): java.lang.RuntimeException: Unable to start receiver com.dexterous.flutterlocalnotifications.ScheduledNotificationReceiver: java.lang.ClassCastException: java.util.ArrayList cannot be cast to byte[]
E/AndroidRuntime(13396): 	at android.app.ActivityThread.handleReceiver(ActivityThread.java:4041)
E/AndroidRuntime(13396): 	at android.app.ActivityThread.access$1400(ActivityThread.java:237)
E/AndroidRuntime(13396): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1924)
E/AndroidRuntime(13396): 	at android.os.Handler.dispatchMessage(Handler.java:106)
E/AndroidRuntime(13396): 	at android.os.Looper.loop(Looper.java:223)
E/AndroidRuntime(13396): 	at android.app.ActivityThread.main(ActivityThread.java:7656)
E/AndroidRuntime(13396): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(13396): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
E/AndroidRuntime(13396): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
E/AndroidRuntime(13396): Caused by: java.lang.ClassCastException: java.util.ArrayList cannot be cast to byte[]
E/AndroidRuntime(13396): 	at com.dexterous.flutterlocalnotifications.FlutterLocalNotificationsPlugin.getBitmapFromSource(FlutterLocalNotificationsPlugin.java:463)
E/AndroidRuntime(13396): 	at com.dexterous.flutterlocalnotifications.FlutterLocalNotificationsPlugin.createNotification(FlutterLocalNotificationsPlugin.java:181)
E/AndroidRuntime(13396): 	at com.dexterous.flutterlocalnotifications.FlutterLocalNotificationsPlugin.showNotification(FlutterLocalNotificationsPlugin.java:794)
E/AndroidRuntime(13396): 	at com.dexterous.flutterlocalnotifications.ScheduledNotificationReceiver.onReceive(ScheduledNotificationReceiver.java:46)
E/AndroidRuntime(13396): 	at android.app.ActivityThread.handleReceiver(ActivityThread.java:4032)
E/AndroidRuntime(13396): 	... 8 more
I/Process (13396): Sending signal. PID: 13396 SIG: 9
Lost connection to device.

```

**Comment by MaikuB** [comment]
@lexxxel finished the review of the Linux PR so will look at this PR next. Made some changes within the Dart code to make use of generics but will need to think more about the areas of concern you've highlighted earlier

**Comment by lexxxel** [comment]
@MaikuB thank you very much 🙇

**Comment by MaikuB** [comment]
No, thank you. Any help around contributing to the plugin is very much appreciated :)

**Comment by vytautas-pranskunas-** [comment]
wooohoo.. i see it is merged. Thanks guys.
@MaikuB when you are going to release new version because i still see it is 7.0.0.
(not sure which one was before but in 7.0.0 there is no exported ByteArrayAndroidBitmap)

**Comment by MaikuB** [comment]
When the build goes green on master as well

**Pull request merged**